### ISSUE TITLE
Documentation CI: unresolved rustdoc links in inference/snake.rs and multi_agent/{messages,joint,mod}.rs

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#19]
## Symptom

The `Documentation` job in CI fails with rustdoc "unresolved link" errors in four files:

- `src/inference/snake.rs`
- `src/multi_agent/messages.rs`
- `src/multi_agent/joint.rs`
- `src/multi_agent/mod.rs`

Likely caused by rustdoc bracket-link syntax (`[Foo]`) referencing types that have been renamed, moved, or aren't in scope from where the doc comment lives.

## Reproduction

```bash
cargo +nightly doc --features training --no-deps --document-private-items 2>&1 | grep -E 'unresolved|warning'
```

## Acceptance

- `cargo +nightly doc --features training --no-deps` produces no `unresolved link` warnings (or, if the project policy is to treat warnings as errors in CI, the Documentation CI job goes green).
- All bracketed type/function references in the four files resolve to real items, either via fully-qualified paths or by adding `use` imports in the appropriate scope (or by removing the bracket if the reference was unintentional).

## Discovery

Flagged by the Loom Judge during re-review of PR #16 ([comment](https://github.com/rjwalters/thrust/pull/16#issuecomment-4455152478)).

**Comment by rjwalters** [comment]
## Curator enhancement

### Verified breakage

Reproduced locally via `cargo +nightly doc --features training --no-deps --document-private-items` and CI uses `RUSTDOCFLAGS="-D warnings" cargo doc --no-deps --all-features` on nightly (`.github/workflows/ci.yml:115-120`), so every `unresolved link` warning is a hard CI failure.

**Total unresolved-link warnings: 47** (out of 190 total rustdoc warnings). The issue title under-counts the affected files — the breakage is wider than four files. The actual file distribution:

| File | Count | Notes |
|---|---|---|
| `src/buffer/rollout/storage.rs` | 7 | `[batch_size]` shape annotations |
| `src/buffer/rollout/sampling.rs` | 5 | `[batch_size]` shape annotations |
| `src/policy/inference.rs` | 8 | `[num_actions]`, `[hidden_dim]`, `[action_dim]`, `[1]`, `[obs_dim]` |
| `src/train/ppo/trainer.rs` | 10 | `[batch_size]` shape annotations (two doc blocks) |
| `src/utils/normalize.rs` | 3 | `[obs_dim]`, `[batch_size][obs_dim]` |
| `src/buffer/rollout/gae.rs` | 2 | `[num_envs]` |
| `src/buffer/rollout.rs` | 1 | `[num_envs]` |
| `src/train/ppo/loss.rs` | 1 | `[num_envs]` |
| `src/policy/universal_inference.rs` | 1 | `[size]` |
| `src/policy/multi_discrete_mlp.rs` | 1 | `[`MlpPolicy`]` — actual type reference, not a shape |
| `src/inference/snake.rs` | 1 | `[num_actions]` |
| `src/multi_agent/mod.rs` | 5 | `[`PolicyLearner`]`, `[`Population`]`, `[`GameSimulator`]`, `[`joint::JointMultiAgentTrainer`]`, `[`joint`]` |
| `src/multi_agent/joint.rs` | 1 | `[`Self::with_adam`]` — method doesn't exist |
| `src/multi_agent/messages.rs` | 2 | `[obs_dim]` shape annotations |

Two distinct root causes:

1. **Shape-annotation false positives (42 of 47).** Doc comments use `[obs_dim]`, `[batch_size]`, `[num_envs]`, etc. as shape notation. Rustdoc parses `[name]` as an intra-doc link to an item named `name` and fails because no such item exists. These are documentation-only; the comments display the intended text fine on a webpage but trip `-D warnings`.

2. **Genuine broken references (5 of 47).** Real type/item names that rustdoc can't resolve from the scope of the doc comment.

---

### Per-warning punch list

Each row below is `file:line` → exact bracket text → fix.

#### Group A: shape-annotation false positives — fix by escaping or backticking

Pattern: turn `[name]` into either `` `[name]` `` (the shape goes inside an inline code span so rustdoc won't treat it as a link) or `\[name\]` (escape). Backticks read better and match the style already used elsewhere in this codebase (e.g. ``` `obs` - Observation vector ```). Recommended fix below is the backtick form.

| File:Line | Current | Fix |
|---|---|---|
| `src/inference/snake.rs:115` | `Action logits [num_actions] and state value` | `` Action logits `[num_actions]` and state value `` |
| `src/multi_agent/messages.rs:17` | `Observation tensor [obs_dim]` | `` Observation tensor `[obs_dim]` `` |
| `src/multi_agent/messages.rs:26` | `Next observation tensor [obs_dim]` | `` Next observation tensor `[obs_dim]` `` |
| `src/policy/inference.rs:132` | `Action logits [num_actions] and state value` | `` Action logits `[num_actions]` and state value `` |
| `src/policy/inference.rs:274` | `Shared layer 1: bias [hidden_dim]` | `` Shared layer 1: bias `[hidden_dim]` `` |
| `src/policy/inference.rs:279` | `Shared layer 2: bias [hidden_dim]` | `` Shared layer 2: bias `[hidden_dim]` `` |
| `src/policy/inference.rs:284` | `Policy head: bias [action_dim]` | `` Policy head: bias `[action_dim]` `` |
| `src/policy/inference.rs:289` | `Value head: bias [1]` | `` Value head: bias `[1]` `` |
| `src/policy/inference.rs:330` | `Observation vector [obs_dim]` | `` Observation vector `[obs_dim]` `` |
| `src/policy/inference.rs:333` | `Action logits [action_dim] and state value` | `` Action logits `[action_dim]` and state value `` |
| `src/policy/inference.rs:380` | `Observation vector [obs_dim]` | `` Observation vector `[obs_dim]` `` |
| `src/policy/universal_inference.rs:305` | `1D tensor [size]` | `` 1D tensor `[size]` `` |
| `src/buffer/rollout/gae.rs:14` | `Value estimates for the final states [num_envs]` | `` ...for the final states `[num_envs]` `` |
| `src/buffer/rollout/gae.rs:139` | `Value estimates for final states [num_envs]` | `` ...for final states `[num_envs]` `` |
| `src/buffer/rollout.rs:39` | `Value estimates for the final states [num_envs]` | `` ...for the final states `[num_envs]` `` |
| `src/buffer/rollout/sampling.rs:94` | `Actions [batch_size]` | `` Actions `[batch_size]` `` |
| `src/buffer/rollout/sampling.rs:97` | `Old log probabilities [batch_size]` | `` Old log probabilities `[batch_size]` `` |
| `src/buffer/rollout/sampling.rs:100` | `Old value estimates [batch_size]` | `` Old value estimates `[batch_size]` `` |
| `src/buffer/rollout/sampling.rs:103` | `Advantages [batch_size]` | `` Advantages `[batch_size]` `` |
| `src/buffer/rollout/sampling.rs:106` | `Returns [batch_size]` | `` Returns `[batch_size]` `` |
| `src/buffer/rollout/storage.rs:210` | `All arrays have shape [batch_size].` | `` All arrays have shape `[batch_size]`. `` |
| `src/buffer/rollout/storage.rs:216` | `Actions taken [batch_size]` | `` Actions taken `[batch_size]` `` |
| `src/buffer/rollout/storage.rs:219` | `Old log probabilities [batch_size]` | `` Old log probabilities `[batch_size]` `` |
| `src/buffer/rollout/storage.rs:222` | `Old value estimates [batch_size]` | `` Old value estimates `[batch_size]` `` |
| `src/buffer/rollout/storage.rs:225` | `Computed advantages [batch_size]` | `` Computed advantages `[batch_size]` `` |
| `src/buffer/rollout/storage.rs:228` | `Computed returns [batch_size]` | `` Computed returns `[batch_size]` `` |
| `src/train/ppo/loss.rs:117` | `Next step value estimates [num_envs]` | `` Next step value estimates `[num_envs]` `` |
| `src/train/ppo/trainer.rs:86` | `Action tensor [batch_size]` | `` Action tensor `[batch_size]` `` |
| `src/train/ppo/trainer.rs:87` | `Old log probabilities [batch_size]` | `` Old log probabilities `[batch_size]` `` |
| `src/train/ppo/trainer.rs:88` | `Old value estimates [batch_size]` | `` Old value estimates `[batch_size]` `` |
| `src/train/ppo/trainer.rs:89` | `Computed advantages [batch_size]` | `` Computed advantages `[batch_size]` `` |
| `src/train/ppo/trainer.rs:90` | `Computed returns [batch_size]` | `` Computed returns `[batch_size]` `` |
| `src/train/ppo/trainer.rs:334` | `Action tensor [batch_size]` | `` Action tensor `[batch_size]` `` |
| `src/train/ppo/trainer.rs:335` | `Old log probabilities [batch_size]` | `` Old log probabilities `[batch_size]` `` |
| `src/train/ppo/trainer.rs:336` | `Old value estimates [batch_size]` | `` Old value estimates `[batch_size]` `` |
| `src/train/ppo/trainer.rs:337` | `Computed advantages [batch_size]` | `` Computed advantages `[batch_size]` `` |
| `src/train/ppo/trainer.rs:338` | `Computed returns [batch_size]` | `` Computed returns `[batch_size]` `` |
| `src/utils/normalize.rs:31` | `Batch of observations [batch_size][obs_dim]` | `` Batch of observations `[batch_size][obs_dim]` `` |
| `src/utils/normalize.rs:90` | `Observations to normalize [obs_dim]` | `` Observations to normalize `[obs_dim]` `` |
| `src/utils/normalize.rs:106` | `Batch of observations [batch_size][obs_dim]` | `` Batch of observations `[batch_size][obs_dim]` `` |

Also the implicit double-bracket cases at `utils/normalize.rs:31`, `utils/normalize.rs:106` only emit one warning each (rustdoc reports the second `[obs_dim]`), but wrapping the entire shape in one set of backticks (as shown) fixes both halves in one edit.

#### Group B: genuine broken references — real items, fix with path or remove

| File:Line | Current bracket text | Root cause | Recommended fix |
|---|---|---|---|
| `src/multi_agent/mod.rs:5` | `` [`PolicyLearner`] `` | This is a crate-root `//!` module doc; the symbol `PolicyLearner` is re-exported on line 42 of the same file but the doc comment is parsed before the re-exports are in scope as plain idents from the module-level doc rendering perspective. Rustdoc resolves names against the *containing module's* scope at the point of the doc, and `PolicyLearner` only exists inside `learner::`. | Use module path: `` [`learner::PolicyLearner`] `` (or `` [`crate::multi_agent::PolicyLearner`] ``). |
| `src/multi_agent/mod.rs:5` | `` [`Population`] `` | Same — `Population` lives in `population::`. | `` [`population::Population`] `` |
| `src/multi_agent/mod.rs:5` | `` [`GameSimulator`] `` | Same — `GameSimulator` lives in `simulator::`. | `` [`simulator::GameSimulator`] `` |
| `src/multi_agent/mod.rs:17` | `` [`joint::JointMultiAgentTrainer`] `` | The `joint` submodule is declared on line 30, *after* the `//!` doc block. Rustdoc cannot see submodules declared below the inner doc comment from within that same inner doc. | Either move the `pub mod` declarations above the `//!` doc (not idiomatic) **or** use the fully-qualified path: `` [`crate::multi_agent::joint::JointMultiAgentTrainer`] ``. |
| `src/multi_agent/mod.rs:27` | `` [`joint`] `` | Same submodule-not-yet-in-scope problem. | `` [`crate::multi_agent::joint`] ``. |
| `src/policy/multi_discrete_mlp.rs:13` | `` [`MlpPolicy`] `` | `MlpPolicy` is not in scope from `multi_discrete_mlp.rs`. Need to check whether it lives at `crate::policy::MlpPolicy` or a submodule. | Investigate with `rg "struct MlpPolicy\|pub use.*MlpPolicy" src/`; fix to `` [`crate::policy::MlpPolicy`] `` (or whichever path actually contains it). If the type was renamed/removed, drop the brackets and reword. |
| `src/multi_agent/joint.rs:339` | `` [`Self::with_adam`] `` | The method `with_adam` does not exist on `JointMultiAgentTrainer` (verified: only `new`, `device`, `collect_rollout`, `update`, plus a `JointRollout::zeros` are defined — see `grep -n "pub fn" src/multi_agent/joint.rs`). The doc comment is *promising* a constructor that was never landed. | Two options, builder's choice: **(a)** delete the entire doc paragraph that promises the helper (the surrounding text on lines 337-340 describing "Use [`Self::with_adam`] for the common case…" — clearest if the helper is not coming back), **or (b)** if a future PR will add `with_adam`, replace the link with prose like ``"a future helper (tracked separately) will provide a fresh Adam optimizer per policy"`` to keep CI green without making a false promise. Filing a follow-up issue for the missing helper would be appropriate either way. |

---

### Implementation guidance

**Recommended approach: a single mechanical pass.**

1. **Group A (42 warnings):** these are all the same pattern. The fastest fix is to wrap the bracketed shape in inline-code backticks. Each is a one-character-each-side edit. A scripted approach is possible but probably over-engineered for 42 lines — straight Edit calls per line are fine.

2. **Group B (5 warnings):** edit `src/multi_agent/mod.rs` lines 5, 17, 27 to use module-qualified paths as shown above. Edit `src/policy/multi_discrete_mlp.rs:13` after confirming the canonical path of `MlpPolicy`. Edit `src/multi_agent/joint.rs:339-340` to remove the broken `[`Self::with_adam`]` reference (preferred: option (a) above, deletion of the promise sentence).

**Why backticks rather than backslash escape:**
- Backticks render the shape as inline code (visually clearer for `[batch_size][obs_dim]`-style annotations).
- Backslash escape `\[batch_size\]` works for rustdoc but the literal `\` shows up in some renderers and looks odd in IDE hover tooltips.
- The codebase already uses backticks around identifier names (e.g. `` `obs` - Observation vector ``), so backticks match the surrounding style.

**Out-of-scope but related:** the original issue lists only 4 files. The builder should fix all 47 in one PR since (a) they share a single root cause for Group A, (b) CI uses `-D warnings` and will keep failing as long as any remain, and (c) splitting this across 12 PRs would be wasteful. Confirm with maintainer before expanding scope if uncertain — but the current issue's acceptance criterion ("Documentation CI job goes green") cannot be met by fixing only 4 files.

### Acceptance criteria

- `RUSTDOCFLAGS="-D warnings" cargo +nightly doc --no-deps --all-features` exits successfully (matches CI command in `.github/workflows/ci.yml:120`).
- Equivalently, `cargo +nightly doc --features training --no-deps --document-private-items 2>&1 | grep -c "unresolved link"` outputs `0`.
- The `Documentation` job in CI goes green on the resulting PR.
- No `///` or `//!` doc comment is removed except for the `Self::with_adam` reference whose target does not exist; all other shape annotations and type references must be preserved (just escaped or path-qualified).

### Test plan

1. **Local rustdoc:** run `RUSTDOCFLAGS="-D warnings" cargo +nightly doc --no-deps --all-features` and confirm exit code 0 and no warnings emitted to stderr.
2. **Sanity-check that shape annotations still render correctly:** run `cargo +nightly doc --features training --no-deps` and open `target/doc/thrust_rl/inference/snake/struct.SnakePolicy.html` (or equivalent) in a browser; the `[num_actions]` text should appear as inline code, not as a broken link rendering.
3. **CI verification:** push to a branch, observe the `docs` job in `.github/workflows/ci.yml` passing.
4. **No regressions:** `cargo +nightly check --all-features` still passes (doc-only changes shouldn't affect this but verifies no stray syntactic issue introduced by escaped brackets).

### Code references

- CI workflow: `/Users/rwalters/GitHub/thrust/.github/workflows/ci.yml:101-120`
- Four originally-cited files: `/Users/rwalters/GitHub/thrust/src/inference/snake.rs:115`, `/Users/rwalters/GitHub/thrust/src/multi_agent/messages.rs:17`, `/Users/rwalters/GitHub/thrust/src/multi_agent/messages.rs:26`, `/Users/rwalters/GitHub/thrust/src/multi_agent/joint.rs:339`, `/Users/rwalters/GitHub/thrust/src/multi_agent/mod.rs:5,17,27`
- Additional files needed for green CI (not in original issue): see Group A table above.

### Open questions for the maintainer

1. Is `with_adam` on `JointMultiAgentTrainer` planned (option b) or abandoned (option a)? Recommend a follow-up issue if planned.
2. Should I open a separate issue to broaden scope, or expand this issue's title to cover all 12 affected files? Suggested updated title: "Documentation CI: 47 unresolved rustdoc links across `policy`, `buffer`, `train`, `multi_agent`, `inference`, `utils`".
